File: 2812-find-the-safest-path-in-a-grid/2812-find-the-safest-path-in-a-grid.cpp
class Solution {
    static constexpr int dirs[4][2] = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};

public:
    int maximumSafenessFactor(vector<vector<int>>& A) {
        if (A[0][0] || A.back().back())
            return 0;

        int n = A.size();
        queue<pair<int, int>> q;

        for (int i = 0; i < n; i++)
            for (int j = 0; j < n; j++)
                if (A[i][j])
                    q.push({i, j});

        while (q.size()) {
            auto [i, j] = q.front();
            q.pop();

            int v = A[i][j];

            for (auto& d : dirs) {
                int x = i + d[0];
                int y = j + d[1];

                if (min(x, y) >= 0 && max(x, y) < n && !A[x][y]) {
                    A[x][y] = v + 1;
                    q.push({x, y});
                }
            }
        }



        priority_queue<tuple<int, int, int>> pq;

        pq.push({A[0][0], 0, 0});

        while (pq.size()) {
            auto [sf, i, j] = pq.top();
            pq.pop();

            if (i == n - 1 && j == n - 1)
                return sf - 1;

            for (auto& d : dirs) {
                int x = i + d[0];
                int y = j + d[1];

                if (min(x, y) >= 0 && max(x, y) < n && A[x][y] > 0) {
                    pq.push({min(sf, A[x][y]), x, y});
                    A[x][y] *= -1;
                }
            }
        }

        return A.back().back() - 1;
    }
};

File: 2812-find-the-safest-path-in-a-grid/2812-find-the-safest-path-in-a-grid.java
class Solution {
    static final int[][] dirs = {{1, 0}, {0, 1}, {-1, 0}, {0, -1}};

    public int maximumSafenessFactor(List<List<Integer>> grid) {
        int n = grid.size();

        if (grid.get(0).get(0) == 1 || grid.get(n - 1).get(n - 1) == 1)
            return 0;

        int[][] A = new int[n][n];

        for (int i = 0; i < n; i++)
            for (int j = 0; j < n; j++)
                A[i][j] = grid.get(i).get(j);

        Queue<int[]> q = new LinkedList<>();

        for (int i = 0; i < n; i++)
            for (int j = 0; j < n; j++)
                if (A[i][j] == 1)
                    q.add(new int[]{i, j});

        while (q.size() > 0) {
            int[] head = q.poll();
            int i = head[0];
            int j = head[1];
            int v = A[i][j];

            for (int[] d : dirs) {
                int x = i + d[0];
                int y = j + d[1];

                if (Math.min(x, y) >= 0 && Math.max(x, y) < n && A[x][y] == 0) {
                    A[x][y] = v + 1;
                    q.add(new int[]{x, y});
                }
            }
        }

        PriorityQueue<int[]> pq = new PriorityQueue<>((a, b) -> b[0] - a[0]);
        pq.add(new int[]{A[0][0], 0, 0});

        while (pq.size() > 0) {
            int[] head = pq.poll();
            int sf = head[0];
            int i = head[1];
            int j = head[2];

            if (i == n - 1 && j == n - 1)
                return sf - 1;

            for (int[] d : dirs) {
                int x = i + d[0];
                int y = j + d[1];

                if (Math.min(x, y) >= 0 && Math.max(x, y) < n && A[x][y] > 0) {
                    pq.add(new int[]{Math.min(sf, A[x][y]), x, y});
                    A[x][y] *= -1;
                }
            }
        }

        return A[n - 1][n - 1] - 1;
    }
}